building_sim_plugins/building_gazebo_plugins/src/crowd_simulator.cpp

```diff
@@ -0,0 +1,499 @@
+  gzdbg << "Gazebo models all loaded! Start simulating..." << std::endl;
+
+}
+
+//============================================
+void CrowdSimulatorPlugin::_UpdateObject(double deltaTime, double deltaSimTime,
+  const crowd_simulator::AgentPtr agentPtr,
+  const gazebo::physics::ModelPtr modelPtr,
+  const crowd_simulator::ModelTypeDatabase::Record* typePtr)
+{
+
+  if (!agentPtr)
+  {
+    gzerr << "Null agentPtr when update Object!" << std::endl;
+    assert(agentPtr);
+  }
+  if (!modelPtr)
+  {
+    gzerr << "Null modelPtr when update Object!" << std::endl;
+    assert(modelPtr);
+  }
+
+  ignition::math::Pose3d pose;
+  if (agentPtr->_external)
+  {
+    //update pose from gazebo to menge
+    pose = modelPtr->WorldPose();
+    this->_crowdSimInterface->UpdateExternalAgent(agentPtr, Convert(pose));
+    return;
+  }
+
+  // only update the external agents
+  if(deltaSimTime - 0.0 < 1e-6) return;
+
+  //update pose from menge to gazebo
+  crowd_simulator::AgentPose3d agent_pose;
+  this->_crowdSimInterface->GetAgentPose(agentPtr, deltaSimTime, agent_pose);
+  pose = Convert(agent_pose); 
+
+  gazebo::physics::ActorPtr actorPtr =
+    boost::dynamic_pointer_cast<gazebo::physics::Actor>(modelPtr);
+
+  auto delta_dist_vector = pose.Pos() - actorPtr->WorldPose().Pos() ;
+  delta_dist_vector.Z(0.0);
+  double deltaDist = delta_dist_vector.Length();
+
+  // _simTimeStep is small, then deltaDist is small, the scriptTime is small than expected.
+  actorPtr->SetScriptTime(
+    actorPtr->ScriptTime() + deltaDist / typePtr->animationSpeed );
+
+  //add on original loaded pose
+  auto animation = actorPtr->SkeletonAnimations().at(typePtr->animation);
+  auto animPose = this->_AnimationRootPose(actorPtr, animation);
+  animPose += Convert(typePtr->pose);
+
+  //update x and y coordinates
+  animPose.Pos().X(pose.Pos().X());
+  animPose.Pos().Y(pose.Pos().Y());
+  animPose.Rot() = pose.Rot() * animPose.Rot();
+
+  actorPtr->SetWorldPose(animPose);
+
+}
+
+
+//============================================
+ignition::math::Pose3d CrowdSimulatorPlugin::_AnimationRootPose(
+  const gazebo::physics::ActorPtr actorPtr,
+  const gazebo::common::SkeletonAnimation* animation)
+{
+
+  auto* rootNode = actorPtr->Mesh()->GetSkeleton()->GetRootNode();
+  auto rootNodeName = rootNode->GetName();
+  if (!animation->HasNode(rootNodeName))
+  {
+    throw std::runtime_error("unable to find root node pose");
+  }
+  //get the animation trans for current time
+  auto animTrans = animation->PoseAt(actorPtr->ScriptTime(), true); //map<string, matrix4d>
+  auto& rootAnimTrans = animTrans[rootNodeName];
+
+  auto scaleTrans = ignition::math::Matrix4d::Identity;
+  auto actorScale = actorPtr->Scale();
+  scaleTrans.Scale(actorScale.X(), actorScale.Y(), actorScale.Z());
+  rootAnimTrans = scaleTrans * rootAnimTrans;
+  return rootAnimTrans.Pose();
+}
+
+
+//============================================
+bool CrowdSimulatorPlugin::_LoadParams(const sdf::ElementPtr& sdf)
```

Comment:
I believe all the sdf parsing could also be moved to the common library, since both ignition and gazebo parse the same `<plugin>` tag

Reply:
Yes, this can be done.